pkg/querier/queryrange/querysharding.go

```diff
@@ -0,0 +1,311 @@
+	if err != nil {
+		return nil, err
+	}
+
+	subtreeFolder, err := astmapper.NewSubtreeFolder(astmapper.JSONCodec)
+	if err != nil {
+		return nil, err
+	}
+
+	strQuery := r.GetQuery()
+	mappedQuery, err := mapQuery(
+		astmapper.NewMultiMapper(
+			shardSummer,
+			subtreeFolder,
+		),
+		strQuery,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	strMappedQuery := mappedQuery.String()
+	level.Debug(ast.logger).Log("msg", "mapped query", "original", strQuery, "mapped", strMappedQuery)
+
+	return ast.next.Do(ctx, r.WithQuery(strMappedQuery))
+
+}
+
+type queryShard struct {
+	getConf func(Request) (chunk.PeriodConfig, error)
+	next    Handler
+	engine  *promql.Engine
+}
+
+func (qs *queryShard) Do(ctx context.Context, r Request) (Response, error) {
+	// since there's no available sharding configuration for this time range,
+	// no astmapping has been performed, so skip this middleware.
+	if _, err := qs.getConf(r); err != nil {
+		return qs.next.Do(ctx, r)
+	}
+
+	queryable := lazyquery.NewLazyQueryable(&DownstreamQueryable{r, qs.next})
+
+	qry, err := qs.engine.NewRangeQuery(
+		queryable,
+		r.GetQuery(),
+		TimeFromMillis(r.GetStart()),
+		TimeFromMillis(r.GetEnd()),
+		time.Duration(r.GetStep())*time.Millisecond,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+	res := qry.Exec(ctx)
+	extracted, err := FromResult(res)
+	if err != nil {
+		return nil, err
+
+	}
+	return &PrometheusResponse{
+		Status: StatusSuccess,
+		Data: PrometheusData{
+			ResultType: string(res.Value.Type()),
+			Result:     extracted,
+		},
+	}, nil
+}
+
+// shardSplitter middleware will only shard appropriate requests that do not extend past the MinShardingLookback interval.
+// This is used to send nonsharded requests to the ingesters in order to not overload them.
+type shardSplitter struct {
+	codec               Codec
+	MinShardingLookback time.Duration    // delimiter for splitting sharded vs non-sharded queries
+	shardingware        Handler          // handler for sharded queries
+	next                Handler          // handler for non-sharded queries
+	now                 func() time.Time // injectable time.Now
+}
+
+func (splitter *shardSplitter) Do(ctx context.Context, r Request) (Response, error) {
+	cutoff := splitter.now().Add(-splitter.MinShardingLookback)
+	sharded, nonsharded := partitionRequest(r, cutoff)
+
+	return splitter.parallel(ctx, sharded, nonsharded)
+
+}
+
+func (splitter *shardSplitter) parallel(ctx context.Context, sharded, nonsharded Request) (Response, error) {
+	if sharded == nil {
+		return splitter.next.Do(ctx, nonsharded)
+	}
+
+	if nonsharded == nil {
+		return splitter.shardingware.Do(ctx, sharded)
+	}
+
+	nonshardCh := make(chan Response, 1)
+	shardCh := make(chan Response, 1)
+	errCh := make(chan error, 2)
+
+	go func() {
+		res, err := splitter.next.Do(ctx, nonsharded)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		nonshardCh <- res
```

Comment:
Should you select on sending the result or context canceled ? I guess if we get cancelled this shouldn't stay blocked on sending.

same goes for line 266.

Reply:
the channels are buffered here so it should be fine